### DISCUSSION TITLE
Deleting a branch after a squash and merge on GitHub also deletes its commits

### DISCUSSION THREAD

**Discussion by nathanael-naeri** [community/community#21689] [New to GitHub]
<p>Hi, I’ve accepted a pull request on GitHub through a squash and merge of the feature branch into the master branch:</p>
<p>v master                                    v master</p>
<p>A                      -&gt;   A - - - - - - - E<br></p>
<p> <code>- B - C - D                </code>- B - C - D</p>
<p>            ^ feature                   ^ feature</p>
<p>I then deleted the feature branch by clicking on the button that says so in the pull request, and was surprised to find out from the repository’s graph that the branch’s commits, not just the branch’s name, had been deleted. The repository’s graph, in Insights &gt; Network on GitHub, now looks like that:</p>
<p>    v master</p>
<p>A - E</p>
<p>A branch, in Git, is a pointer to a commit, or so I’ve read countless times. I expected, and kinda wanted, to only delete this pointer, the branch’s name, leaving out the commits:</p>
<p>                v master</p>
<p>A - - - - - - - E<br></p>
<p> `- B - C - D</p>
<p>Unreacheable commits, granted, but still present, and accessible through their hash.</p>
<p>Is this behavior a bug or a feature of GitHub? If it is a feature, where else should I be cautious that the term “branch” is used to mean “the branch’s commits”? Or am I simply wrong in my understanding of what a branch means in Git?</p>

**Comment by mpboom** [comment, marked as answer]
<pre><code>My history now looked like that:
</code></pre>
<p>What history view do you mean by “My history”?</p>
<p>Anyway, I think this is intended behaviour. By squashing you push all commits of that branch into a single commit, then apply that commit to the master branch. I also believe that with “Delete this branch” GitHub  <strong>does</strong> mean “Delete this branch’s commits”. Actually, I’ve never considered a branch as just a pointer to a commit, but that might as well be true (I am not a Git wizard :wink: ).</p>
<pre><code>where else should I be cautious that the term "branch" is used to mean "the branch's commits
</code></pre>
<p>I would be cautious everywhere about that! Again, at least from my experience, a “branch” mostly never is considered as a pointer but more as a “collection” of commits that originate from a certain point in another branch. Altough, again, that thought may be misplaced, I’d definitely be aware of the differing opinions about what a branch is.</p>

**Comment by nathanael-naeri** [comment]
<blockquote>
<p><a class="mention" href="/mpboom">@mpboom</a> wrote:</p>
</blockquote>
<blockquote>
<p>What history view do you mean by “My history”?</p>
</blockquote>
<p>I meant my repository’s network graph, shown in the Insights tab of the repository on GitHub. I’ve edited my post to clarify.</p>
<blockquote>
<p>I also believe that with “Delete this branch” GitHub  <strong>does</strong> mean “Delete this branch’s commits”.</p>
</blockquote>
<p>It appears so indeed. That is not the case on GitLab, which I was using previously. In my experience of GitLab, deleting a branch through the web interface deletes the branch in the Git sense, that is to say the branch pointer, not the commits.</p>
<blockquote>
<p>Actually, I’ve never considered a branch as just a pointer to a commit, but that might as well be true (I am not a Git wizard :wink: ).</p>
</blockquote>
<p>Yes I’m pretty sure that is the case. I’ve read it countless times, and that seems to match how Git behaves: git branch -d branchname, for instance, simply deletes the branch pointer branchname which points to the commit at the tip of the branch, but that commit and its ancestors are still present in the repository, and accessible through their hash, e.g. git log hash-of-tip-commit.</p>
<blockquote>
<p>I would be cautious everywhere about that! Again, at least from my experience, a “branch” mostly never is considered as a pointer but more as a “collection” of commits that originate from a certain point in another branch. Altough, again, that thought may be misplaced, I’d definitely be aware of the differing opinions about what a branch is.</p>
</blockquote>
<p>I’ve just accepted a pull request, through the rebase and merge strategy this time. Then I clicked on “Delete this branch”, and found out that in this case too, the branch’s commits, not just the branch’s name, where deleted from the repository, according to the repository’s graph. So I will heed your advice!</p>
